src/Data/Document.hs

```diff
@@ -0,0 +1,89 @@
+  Document (Document),
+  documentContent,
+  documentCreatedAt,
+  documentDomain,
+  insertDocument,
+  findDocuments,
+  countDocuments
+  ) where
+
+import           Control.Monad.IO.Class
+import           Data.Aeson             as AESON
+import           Data.Bson              as BSON hiding (Document)
+import           Data.HashMap.Strict    as Map (fromList, toList)
+import           Data.Maybe             (catMaybes)
+import           Data.Scientific
+import           Data.Time
+import           Data.Vector            as Vector (fromList, toList)
+import           Database.MongoDB       as MongoDB hiding (Document)
+
+data Document = Document {
+  documentCreatedAt :: UTCTime,
+  documentDomain    :: String,
+  documentContent   :: AESON.Value
+  } deriving (Eq, Show)
+
+------------------------------------------------------------------------------------------------------------------------
+-- MongoDB queries
+------------------------------------------------------------------------------------------------------------------------
+
+insertDocument :: MonadIO m => Collection -> Document -> Action m MongoDB.Value
+insertDocument col doc = insert col (documentToObj doc)
+
+findDocuments :: MonadIO m => Collection -> String ->  Action m [Document]
+findDocuments col domain = do
+  cursor  <- find (select  ["domain" =: domain] col) { sort = ["date" =: (1 :: Int)] }
+  objects <- rest cursor
+  -- Filter out invalid documents. It'd probably be better to log something here.
+  return . catMaybes $ objToDocument <$> objects
+
+countDocuments :: MonadIO m => Collection -> String ->  Action m Int
+countDocuments col domain = count (select  ["domain" =: domain] col)
+
+------------------------------------------------------------------------------------------------------------------------
+-- BSON.Val instance
+------------------------------------------------------------------------------------------------------------------------
+
+documentToObj :: Document -> [Field]
+documentToObj (Document date domain content) = [
+    "date"    := val date,
+    "domain"  := val domain,
+    "content" := jsonToBSON content
+    ]
+
+objToDocument :: [Field] -> Maybe Document
+objToDocument doc = do
+  date    <- doc !? "date"
+  domain  <- doc !? "domain"
+  content <- bsonToJSON $ valueAt "content" doc
+  return $ Document date domain content
+
+instance BSON.Val Document where
+  val = val . documentToObj
+
+  cast' (BSON.Doc doc) = objToDocument doc
+  cast' _              = Nothing
+
+jsonToBSON :: AESON.Value -> BSON.Value
+jsonToBSON (AESON.Object obj) = BSON.Doc $ tupleToField <$> Map.toList obj
+    where tupleToField (k, v) = k := jsonToBSON v
+jsonToBSON (AESON.Number num) = either BSON.Float BSON.Int64 (floatingOrInteger num)
+jsonToBSON (AESON.Array arr)  = BSON.Array . Vector.toList $ jsonToBSON <$> arr
+jsonToBSON (AESON.String str) = BSON.String str
+jsonToBSON (AESON.Bool bool)  = BSON.Bool bool
+jsonToBSON AESON.Null         = BSON.Null
```

Comment:
I usually find that long pattern matches like this are cleaner and less repetitive with a `case`:

```haskell
jsonToBSON :: AESON.Value -> BSON.Value
jsonToBSON val =
  case val of
    AESON.Object obj -> ...
    AWESON.Number num -> ...
    ...
```

Ymmv

Reply:
Should be fixed in 24e7bde2bcc268d64f3c32bbab00daa4b263af27.